### PR TITLE
generic proxy: added drain support to generic proxy to doing graceful closes on connections when possible

### LINKED ISSUE #24186: generic proxy: drain support
Add drain support to the generic proxy.

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: generic proxy: added drain support to generic proxy to doing graceful closes on connections when possible
Additional Description:

Drain support for generic proxy. To close #24186.

Risk Level: Low. Contrib update.
Testing: unit.
Docs Changes: n/a.
Release Notes: added.
Platform Specific Features: n/a.

### CODE REVIEW

File: contrib/generic_proxy/filters/network/source/proxy.cc [hunk 1]
```diff
@@ -163,6 +164,17 @@ void Filter::resetStreamsForUnexpectedError() {
...
 }
 
+void Filter::mayBeDrainClose() {
+  if (drain_decision_.drainClose() && active_streams_.empty()) {
+    onDrainCloseAndNoActiveStreams();
+  }
+}
+
+// Default implementation for connection draining.
+void Filter::onDrainCloseAndNoActiveStreams() {
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply contrib/generic_proxy/filters/network/source/proxy.cc:174]: It's only a place for the protocol who wants more flexible drain policy to implement its own drain policy.

[the remark this answers is not in the document]
Comment (soulxu) [reply contrib/generic_proxy/filters/network/source/proxy.cc:174]: emm... I thought this should be handled by each stream.

[the remark this answers is not in the document]
Comment (soulxu) [reply contrib/generic_proxy/filters/network/source/proxy.cc:174]: discuss offline with @wbpcode, the stream should check the drainDecision() and then implement the graceful close. For the connection, this is the place to do a graceful close if the protocol needs it.

**Review by soulxu — APPROVED** [review]
lgtm, thanks

**Review by zuercher — APPROVED** [review]

**Pull request merged**